lib/valueflow.cpp

```diff
@@ -1079,6 +1071,58 @@ static void valueFlowArray(TokenList *tokenlist)
     }
 }
 
+static bool isNonZero(const Token *tok)
+{
+    return tok && (!tok->hasKnownIntValue() || tok->values().front().intvalue != 0);
+}
+
+static bool withNonZeroOp(const Token *tok)
+{
+    if (!tok)
+        return false;
+    if (!tok->astParent())
+        return false;
+    if (tok->astParent()->astOperand1() != tok && isNonZero(tok->astParent()->astOperand1()))
+        return true;
+    if (tok->astParent()->astOperand2() != tok && isNonZero(tok->astParent()->astOperand2()))
+        return true;
+    return false;
+}
+
+static void valueFlowArrayBool(TokenList *tokenlist)
+{
+    for (Token *tok = tokenlist->front(); tok; tok = tok->next()) {
+        if (tok->hasKnownIntValue())
+            continue;
+        const Variable *var = nullptr;
+        bool known = false;
+        std::list<ValueFlow::Value>::const_iterator val =
+            std::find_if(tok->values().begin(), tok->values().end(), std::mem_fn(&ValueFlow::Value::isTokValue));
+        if (val == tok->values().end()) {
+            var = tok->variable();
+            known = true;
+        } else {
+            var = val->tokvalue->variable();
+            known = val->isKnown();
+        }
+        if (!var)
+            continue;
+        if (!var->isArray() || var->isArgument() || var->isStlType())
+            continue;
+        if (withNonZeroOp(tok) && Token::Match(tok->astParent(), "%comp%"))
+            continue;
+        // TODO: Check for function argument
+        if (astIsIntegral(tok, false) || Token::Match(tok->astParent(), "&&|%oror%|%comp%") ||
+            (astIsIntegral(tok->astParent(), false) && !Token::Match(tok->astParent(), "(|%name%")) ||
```

Comment:
I am somewhat skeptic. If parent is a integer that usually mean that a pointer-to-integer conversion has been used, doesn't it?

Example code:

    int f() {
      int buf[10];
      int x = buf;
      return x;
    }

Now `ValueFlow` says that `x` will be `1`.

Reply:
I was thinking of trying to catch cases such as this:

```cpp
void f() {
  int buf[10];
  int x = buf;
  if(x) {}
}
```

But I guess using `astIsIntegral` can lead to FPs. I changed it to `astIsBool` for now, and we fix the FNs in the future.